### PR TITLE
Add API to determine what capabilities the runtime supports

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/66620

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Working on tests now.

**Comment by CyrusNajmabadi** [comment]
Tests added.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NativeIntegerTests.cs
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NativeIntegerTests.cs:11]: Seems unnecessary

File: src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb
Comment (jcouv) [inline src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb:3151]: It doesn't seem obvious that we should return false here. 
The alternative is to accurately report what the runtime/TFM supports, regardless of whether the language cares.
I'm curious what others think.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteraiton 5) with a design question for API on VB compilation

File: src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb
Comment (CyrusNajmabadi) [reply src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb:3151]: Good question.  @AlekseyTs what's your thinking here?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NativeIntegerTests.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NativeIntegerTests.cs]: indeed.

File: src/Compilers/CSharp/Test/Semantic/Semantics/InheritanceBindingTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/InheritanceBindingTests.cs:5064]: >Diagnostic(ErrorCode.WRN_MultipleRuntimeOverrideMatches, "Method").WithArguments("Base<int>.Method(int, ref int)", "Derived.Method(int, ref int)").WithLocation(5, 25)); [](http://example.com/codeflow?start=20&length=168)

Please restore original formatting

File: src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb
Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb:3151]: > The alternative is to accurately report what the runtime/TFM supports, regardless of whether the language cares.

I think I am leaning towards this approach as well. It feels that whether language supports something and whether runtime supports something are orthogonal things. For example, runtime might support some feature, but it might be unavailable due to target language version. Another example is the fact that DIM feature is supported at different levels VB vs. C#. So, the IDE will likely need to take several different aspects into account (runtime support level, language version, language VB vs. C#, etc.) while deciding if it is appropriate to make certain suggestions. Also, it might be confusing to users when for the same runtime VB gives a different answer by comparison to C#.

Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb:3151]: Alternatively we can rename the API to reflect the dependency (it is not purely about runtime) and clearly document it.

Comment (CyrusNajmabadi) [reply src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb:3151]: > I think I am leaning towards this approach as well.

Sounds good.  Let me try moving to that approach.  Will ping when ready.

File: src/Compilers/Core/Portable/RuntimeCapability.cs
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/RuntimeCapability.cs:74]: the logic of these was pulled from the C# impl.  I'm not sure why some are different (like why NumericIntPtr explicitly checks .CorLibrary, but the rest do not).

File: src/Compilers/CSharp/Portable/Symbols/AssemblySymbol.cs
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Symbols/AssemblySymbol.cs:415]: the logic here is pulled into C# and VB separately as the two languages were doing the check slightly differently in the previous code i was updating.  so this keeps things preserved in that regard.

File: src/Compilers/CSharp/Portable/Symbols/AssemblySymbol.cs
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Symbols/AssemblySymbol.cs:421]: existing properties now just defer to helper.

File: src/Compilers/Core/Portable/RuntimeCapability.cs
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/RuntimeCapability.cs:85]: vb and C# are slightly different how they test if something is a static class.  hence why we defer to the language to make the determination here.

**Comment by CyrusNajmabadi** [comment]
Ok.  Have updated "to accurately report what the runtime/TFM supports, regardless of whether the language cares."  LMK what you think of this @AlekseyTs @jcouv .

File: src/Compilers/Core/Portable/RuntimeCapability.cs
Comment (jcouv) [reply src/Compilers/Core/Portable/RuntimeCapability.cs:78]: The reason NumericIntPtr from the others is that it is involved in metadata decoding. This flag affects how we read `System.IntPtr`. 
There's a bug tracking that it is possible to have null CorLibrary, so it needed special handling. Once that bug is fixed, the logic should be simple like the other cases.

File: src/Compilers/CSharp/Portable/Symbols/AssemblySymbol.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/AssemblySymbol.cs:414]: nit: Why is this method a private method implementation, but the other similar ones are not?

File: src/Compilers/Core/Portable/RuntimeCapability.cs
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/RuntimeCapability.cs:78]: All the others end up calling into GetSpecialTypeMember, which does thsi:

```c#
        internal virtual Symbol GetSpecialTypeMember(SpecialMember member)
        {
            return CorLibrary.GetDeclaredSpecialTypeMember(member);
        }
```

So it looks like the others will just crash if you have a null CorLibrary. :)

File: src/Compilers/VisualBasic/Portable/Symbols/AssemblySymbol.vb
Comment (jcouv) [inline src/Compilers/VisualBasic/Portable/Symbols/AssemblySymbol.vb:350]: >namedType.IsMetadataAbstract [](http://example.com/codeflow?start=51&length=28)

`IsMetadataAbstract` does `Me.IsMustInherit OrElse Me.IsInterface`. We could inline the `IsMustInherit` logic here and drop the `IsInterface` (redundant since we're only interested in classes here).

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 14)

**Comment by jcouv** [comment]
@AlekseyTs for another look. Thanks

File: src/Compilers/Core/Portable/RuntimeCapability.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/RuntimeCapability.cs:69]: >// CorLibrary should never be null, but that invariant is broken in some cases for MissingAssemblySymbol. [](http://example.com/codeflow?start=20&length=105)

I think this issue is specific to EnC scenarios only, caused by the fact how EnC creates symbols. Other consumers shouldn't be able to run into this problem. In addition, it looks very strange that we are protecting only this case.

File: src/Compilers/Core/Portable/RuntimeCapability.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/RuntimeCapability.cs:51]: >RuntimeCapabilityHelpers [](http://example.com/codeflow?start=26&length=24)

I do not like this refactoring. It creates to much unnecessary churn. My reference would be to keep each language doing its own work and not adding more APIs to ```IAssemblySymbolInternal```. When looking at the change, I would prefer it to be obvious (because there are no code changes) that internal behavior of the compiler is not changed. We are only adding a public API that that calls into existing internal implementation. Obviously, for VB we will need to add more internal implementation.

Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/RuntimeCapability.cs:51]: I get that. Though my hope would be that our extensive tests validate that the behavior of the compiler is unchanged. 

The virtue of the refactoring here is that we truly are sharing the same logic between c# and vb. So there's no risk of getting it wrong between them. For example, if one of the checks needs to validate a particular attribute is present, having the logic be unified means we do that properly for both languages (as opposed to potentially forgetting it for one of them). 

Also, since we did say we wanted to be accurate here, regardless of the language, it really does help (and bake a lot of sense) to unify the impl to ensure that language cannot affect this. 

I can change this. But I think it will make it much easier to introduce mistakes in the future.

Comment (AlekseyTs) [reply src/Compilers/Core/Portable/RuntimeCapability.cs:51]: I understand the motivation, but my opinion remains the same. Among other things, I especially dislike extending the ```IAssemblySymbolInternal``` in this fasion.

> Though my hope would be that our extensive tests validate that the behavior of the compiler is unchanged.

No change is the best guarantee for no behavior change.

Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/RuntimeCapability.cs:51]: Ok. Just to confirm. You're stating this as more than a preference, and as an explicit change request for the pr?  

If it's just a preference, I'd prefer to keep as is. If it's an explicit request to change it, I'll make the change. 

Let me know which case this is and I'll follow-up accordingly. Thanks!

Comment (AlekseyTs) [reply src/Compilers/Core/Portable/RuntimeCapability.cs:51]: > Just to confirm. You're stating this as more than a preference, and as an explicit change request for the pr?

I will not sign-off on the change as is.

**Comment by CyrusNajmabadi** [comment]
Ok.  Logic now duplicated to VB.  Comments added to both sides to help maintainers know to keep both copies in sync.

File: src/Compilers/Core/Portable/RuntimeCapability.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/RuntimeCapability.cs:6]: Are these usings necessary?

File: src/Compilers/CSharp/Portable/Symbols/AssemblySymbol.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/AssemblySymbol.cs:433]: >return false; [](http://example.com/codeflow?start=12&length=13)

It is probably worth documenting that ```false``` is returned for an unrecognized input (rather than throw).

File: src/Compilers/CSharp/Test/Semantic/Semantics/InheritanceBindingTests.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/InheritanceBindingTests.cs:5064]: Please restore original formatting

**Comment by AlekseyTs** [comment]
Done with review pass (commit 17)

File: src/Compilers/CSharp/Portable/Symbols/AssemblySymbol.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Symbols/AssemblySymbol.cs:433]: will do.

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 20)

**Comment by AlekseyTs** [comment]
@CyrusNajmabadi Please squash commits while merging.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 20)

**Pull request merged**
